dex/networks/eth/contracts/ETHSwapV1.sol

```diff
@@ -0,0 +1,231 @@
+            r.secret = record;
+        } else {
+            r.step = Step.Filled;
+            r.blockNumber =  blockNum;
+        }
+        return r;
+    }
+
+    // initiate initiates an array of Vectors.
+    function initiate(Vector[] calldata contracts)
+        public
+        payable
+        senderIsOrigin()
+    {
+        uint initVal = 0;
+        for (uint i = 0; i < contracts.length; i++) {
+            Vector calldata v = contracts[i];
+
+            require(v.value > 0, "0 val");
+            require(v.refundTimestamp > 0, "0 refundTimestamp");
+
+            bytes32 k = contractKey(v);
+            bytes32 record = swaps[k];
+            require(record == bytes32(0), "swap not empty");
+
+            record = bytes32(block.number);
+            require(!secretValidates(record, v.secretHash), "hash collision");
+
+            swaps[k] = record;
+
+            initVal += v.value;
+        }
+
+        require(initVal == msg.value, "bad val");
+    }
+
+    // isRedeemable returns whether or not a swap identified by secretHash
+    // can be redeemed using secret. isRedeemable DOES NOT check if the caller
+    // is the participant in the vector.
+    function isRedeemable(Vector calldata v)
+        public
+        view
+        returns (bool)
+    {
+        (, bytes32 record, uint256 blockNum) = retrieveStatus(v);
+        return blockNum != 0 && !secretValidates(record, v.secretHash);
+    }
+
+    // redeem redeems a Vector. It checks that the sender is not a contract,
+    // and that the secret hash hashes to secretHash. msg.value is tranfered
+    // from ETHSwap to the sender.
+    //
+    // To prevent reentry attack, it is very important to check the state of the
+    // contract first, and change the state before proceeding to send. That way,
+    // the nested attacking function will throw upon trying to call redeem a
+    // second time. Currently, reentry is also not possible because contracts
+    // cannot use this contract.
+    function redeem(Redemption[] calldata redemptions)
+        public
+        senderIsOrigin()
+    {
+        uint amountToRedeem = 0;
+        for (uint i = 0; i < redemptions.length; i++) {
+            Redemption calldata r = redemptions[i];
+
+            require(r.v.participant == msg.sender, "not authed");
+
+            (bytes32 k, bytes32 record, uint256 blockNum) = retrieveStatus(r.v);
+
+            // To be redeemable, the record needs to represent a valid block
+            // number.
+            require(blockNum > 0 && blockNum < block.number, "unfilled swap");
+
+            // Can't already be redeemed.
+            require(!secretValidates(record, r.v.secretHash), "already redeemed");
+
+            // Are they presenting the correct secret?
+            require(secretValidates(r.secret, r.v.secretHash), "invalid secret");
+
+            swaps[k] = r.secret;
+
+
+            // DRAFT TODO: NOOOOOO! This doesn't account for decimals. This is
```

Comment:
This `TODO` should have been deleted. I was initially passing in value as a `uint64` and multiplying by `1 gwei` internally. That works for eth, but not for some tokens, so I switched to `uint256` (for both contracts to maintain API compatibility). This does incur a slight bump in gas because there's more call data blocks, but they're not stored so it's not a lot.